### ISSUE TITLE
vmselect: support splitting long time range query by smaller intervals

### ISSUE THREAD

**Issue by Haleygo** [VictoriaMetrics/VictoriaMetrics#5940]
### Is your feature request related to a problem? Please describe

In vmselect, there are [limits](https://docs.victoriametrics.com/cluster-victoriametrics/#resource-usage-limits) like `-search.maxSamplesPerSeries` and `-search.maxSamplesPerQuery` to protect it from using excessive memory and cpu for heavy queries. And it works pretty well for most of the cases, since vmselect can scale vertically and has good performace.
But sometimes, users still want to perform big query without modifying above limits since there is no enough resources, so the idea of spliting a big query into multiple small ones comes

### Describe the solution you'd like

I think there are several options:

1. add parameter `-search.queryRangeSplitInterval` in vmselect
Once getting range query request bigger than `-search.queryRangeSplitInterval`, vmselect will do multiple small queries instead of one, merge the results and return.

- pros: no extra component required.
- cons: there will be only one vmselect involved and could be pretty slow to process big queries.

2. support it in multi-level vmselect setup
vmselect nodes can be queried by other vmselect nodes if they run with `-clusternativeListenAddr` command-line flag, so the top level vmselect can be responsible to split the big request, distribute small requests to second-level vmselect pool, and merge the results.

- pros: make good use of the pool of vmselects.
- cons: now vmselect doesn't know if `-storageNode` is vmstorage or vmselect, need to tell top-level vmselect the `-storageNode` are vmselect as well.

3. add an extra component like query frontend
Like query frontend in [thanos](https://thanos.io/tip/components/query-frontend.md/#query-frontend).

- pros: easy to understand and use.
- cons: might need to add other features like caching.

**Comment by Haleygo** [comment]
cc @hagen1778 @valyala

**Comment by Haleygo** [comment]
related to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5072

**Comment by hagen1778** [comment]
> support it in multi-level vmselect setup
vmselect nodes can be queried by other vmselect nodes if they run with -clusternativeListenAddr command-line flag, so the top level vmselect can be responsible to split the big request, distribute small requests to second-level vmselect pool, and merge the results.

It is not that straightforward, I think. vmselect peers should do pre-processing, otherwise there is no much benefit in this splitting as vmselect peers will just proxy vmstorage responses to the caller. This is no different to just querying vmstorages directly as it is right now.
So in order for this to work, query splitting should work like the following:
1. Initial vmselect gets the query `sum(rate(foo))` on range of 3d.
2. It splits this query in 3 queries, same expr but ranges now() - now() : 1d, now()-1d : now()-2d, now()-2d : now()-3d
3. It sends 3 queries to other vmselects and waits for response
4. Once all responses received, initial vmselect merges response into one on now() : now()-3d range and returns to the user.
In this way, each vmselect spend its resource on compute and returned pre-aggregates to the caller to finalize the response.

This is the most optimal way because:
1. It allows utilize compute resources of vmselect peers
2. It works perfectly for data with high churn rate, as each time range could cover only a fraction of time series
3. It would perfectly utilize caches of peers, especially if we use consistent hashing to choose which peer needs to process which query

But it is also the most complex:
1. vmselect needs to know how to split queries. For example, `sort(foo)` query can't be sharded
2. vmselect needs to know how to finalize pre-aggregates
3. vmselect needs to know about other peers

**Comment by ccope** [comment]
Is there a way for vmselect to estimate how large the sample set is so it can dynamically set the amount of data each subquery fetches to stay within the memory limits?

For example, let's say I want to do an avg_over_time daily rollup over the past year, and there are 500 series matching my label filters. How many subqueries does this need? Could a single vmselect instance handle this by just paginating the results and iterating over each chunk of data?

**Comment by hagen1778** [comment]
> Is there a way for vmselect to estimate how large the sample set is so it can dynamically set the amount of data each subquery fetches to stay within the memory limits?

No, it would require "asking" vmstorage to do a search lookup, locate data blocks on disk, read block headers and respond back to the vmselect. These actions cover about 60% of the expenses on a regular query.

> Could a single vmselect instance handle this by just paginating the results and iterating over each chunk of data?

What's the purpose of the pagination? If for saving memory - all queries are already processed in a streaming fashion, which is better then pagination.

For sharding it also doesnt make sense. To shard the query optimally we need to split the load evenly among all available vmselects. If you have 4 vmselects and you query 360d time range with 1d step, then query needs to be sharded by 4 parts: 0-90, 90-180, 180-270 and 270-360d. Where each vmselect will process its 90d of data.

**Comment by ccope** [comment]
@hagen1778 The purpose/goal behind my question is that I want to be able to issue queries over long historical time ranges without having vmselect overflow its cache storage volume or running out of memory.

**Comment by hagen1778** [comment]
> @hagen1778 The purpose/goal behind my question is that I want to be able to issue queries over long historical time ranges without having vmselect overflow its cache storage volume or running out of memory.

Is this what you observe now? It is unlikely that vmselect could overflow its cache storage, is this storage is persisted on local filesystem - vmselect doesn't hold it in memory. 
The memory usage of vmselect depends on the amount of selected unique time series and the function you apply to them. For example, memory usage of `sort(max(my_metric) by (pod))` expression will depend on the cardinality of `my_metric` because vmselect will have to put all series in memory in order to group them by label and sort by value. Pagination will not help here, because queries like this can't be calculated on the fraction of data.

**Comment by ccope** [comment]
We have seen the cache storage volume attached to our vmselect kubernetes pods run out of space for a query when it pulls in more than 6 months of data. It works for smaller time ranges. I can try to override the cache size but I feel like interactive queries like this one shouldn't even be cached, and repeatedly scaling up the cache volume (as the dataset grows over time) for this one query feels wrong.

**Comment by hagen1778** [comment]
> We have seen the cache storage volume attached to our vmselect kubernetes pods run out of space for a query when it pulls in more than 6 months of data.

What was the size of attached volume?

> I feel like interactive queries like this one shouldn't even be cached

This is not a traditional cache like people are used to think of. Let me clarify the life of the query:
1. vmselect receives the query and parses it. 
2. It then applies modifications to the query, like filter pushdowns to improve the search speed.
3. It checks if there is a rollup cache available for this query. Please note, rollup cache contains not raw data, but processed data returned to user. Like the data you see in Grafana. So a normal query contains a few KiBs of cached data.
4. If no cache available vmselect sends concurrently queries to all vmstorage nodes and awaits for response
5. The response from vmstorage nodes is then in a streaming fashion saved to the disk. The purpose of this action is to:
5.1 receive response **as fast as possible** and release resources occupied in vmstorage nodes
5.2 avoid storing the response in vmselect memory. Instead, process it in streaming fashion by reading from disk when data is needed.
6. Response processing is happening concurrently by N workers in vmselect process. Once all workers are done, response is finalized and streamed back to the user.

As you see, the crucial point here is p5 - release resources as fast as possible. Introducing pagination, as I understand your proposal, would require vmselect to ask for a portion of data, iterate over it, ask for another, and so on. This could save you some disk space, yes. But it will significantly degrade the overall experience due to the following reasons:
1. vmstorage nodes will have to keep resources occupied while vmselect is iterating over chunks of data
2. instead of streaming data with full speed vmselect will have to do many small requests one-by-one, increasing read latency

**Comment by ccope** [comment]
> What was the size of attached volume?

8GB

>     5. The response from vmstorage nodes is then in a streaming fashion saved to the disk. The purpose of this action is to:
>        5.1 receive response **as fast as possible** and release resources occupied in vmstorage nodes
>        5.2 avoid storing the response in vmselect memory. Instead, process it in streaming fashion by reading from disk when data is needed.

The problem is that queries with a long time horizon may return an unbounded amount of data. At some point you want to chunk it so you can fit it into the resources available for the vmselect nodes, as well as start processing while still fetching data.

>     1. vmstorage nodes will have to keep resources occupied while vmselect is iterating over chunks of data

I don't know if vmstorage nodes have any kind of query cache in them, but as long as the lookups for which data to fetch aren't incurring additional processing, it should be returning the same amount of data. And this would only happen for very large queries to begin with.

**Comment by hagen1778** [comment]
> The problem is that queries with a long time horizon may return an unbounded amount of data. At some point you want to chunk it so you can fit it into the resources available for the vmselect nodes, as well as start processing while still fetching data.

If vmselect would shard such a query across other vmselects (peers), it would still consume the same amount of disk space for temp caches. So effectively, the disk usage is not reduced, it just gets distributed.
I don't think if complicating query logic worth that, especially if this problem can be solved by allocating additional disk space to vmselect.
